File: windows/ES_R_DevKit_Windows/MEME_Academic_Sample/Services/DataPersistenceService.cs
using System.Globalization;
using System.Text;
using MEMELib_Academic;

namespace MEME_Academic_Sample.Services;

/// <summary>
/// CSV のヘッダ生成・行整形・バッファ保存を担当する。Mac 版 DataPersistenceService の移植。
/// 1 行ごとに open/close すると 100Hz に追いつかないため、一定件数たまってから書き出す。
/// </summary>
public sealed class DataPersistenceService : IDisposable
{
    private const string DateFormat = "yyyy/MM/dd HH:mm:ss.ff";

    private readonly Lock _gate = new();
    private readonly List<string> _pendingRows = [];

    private string? _directory;
    private string? _fileName;
    private string? _header;
    private int _flushThreshold = 100;

    /// <summary>書き出し中の CSV のパス。まだ 1 件も書き出していなければ null。</summary>
    public string? CurrentFilePath { get; private set; }

    /// <summary>整形済みの 1 行が確定するたびに発火する。TCP 出力へ横流しするために使う。</summary>
    public event Action<string>? RowFormatted;

    /// <summary>計測開始。実ファイルは最初のフラッシュ時に作る(空ファイルを残さないため)。</summary>
    public void Begin(string directory, string macAddress, string header, MEMEQuality quality)
    {
        lock (_gate)
        {
            _pendingRows.Clear();
            _directory = directory;
            _header = header;
            // ファイル名の日時も UTC(DATE 列・Mac 版・Android 版と揃える)。
            _fileName = $"{macAddress}_{DateTime.UtcNow:yyyyMMddHHmmss}.csv";
            _flushThreshold = Math.Max(100 / Math.Max((int)quality, 1), 1);
            CurrentFilePath = null;
        }
    }

    public void Append(AcademicData data, int packetCount, bool freeMarking)
    {
        var row = FormatRow(data, packetCount, freeMarking);
        RowFormatted?.Invoke(row);

        lock (_gate)
        {
            if (_fileName is null)
            {
                return;
            }

            _pendingRows.Add(row);
            if (_pendingRows.Count >= _flushThreshold)
            {
                FlushCore();
            }
        }
    }

    /// <summary>計測停止。残りを書き出してファイルを確定する。</summary>
    public void End()
    {
        lock (_gate)
        {
            FlushCore();
            _fileName = null;
            _header = null;
            _directory = null;
        }
    }

    private void FlushCore()
    {
        if (_pendingRows.Count == 0 || _directory is null || _fileName is null)
        {
            return;
        }

        try
        {
            Directory.CreateDirectory(_directory);
            // ファイル名は秒までしか持たないため、同じ秒に始め直すと衝突しうる。
            // 既存ファイルへ追記すると別セッションが 1 つの CSV に混ざるので、
            // 初回書き出しのときだけ空いている名前を選ぶ。
            var path = CurrentFilePath ?? ResolveUniquePath(Path.Combine(_directory, _fileName));
            var buffer = new StringBuilder();
            if (CurrentFilePath is null)
            {
                buffer.Append(_header);
            }

            foreach (var row in _pendingRows)
            {
                buffer.Append(row).Append("\r\n");
            }

            File.AppendAllText(path, buffer.ToString(), new UTF8Encoding(false));
            CurrentFilePath = path;
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            // 書けなくても計測とチャートは続ける。次のフラッシュで再試行される。
            return;
        }

        _pendingRows.Clear();
    }

    /// <summary>同名のファイルがあれば `_2`, `_3` … を足して空いている名前を返す。</summary>
    private static string ResolveUniquePath(string path)
    {
        if (!File.Exists(path))
        {
            return path;
        }

        var directory = Path.GetDirectoryName(path) ?? string.Empty;
        var name = Path.GetFileNameWithoutExtension(path);
        var extension = Path.GetExtension(path);
        for (var suffix = 2; suffix < 1000; suffix++)
        {
            var candidate = Path.Combine(directory, $"{name}_{suffix}{extension}");
            if (!File.Exists(candidate))
            {
                return candidate;
            }
        }

        return path;
    }

    /// <summary>計測パラメータから CSV ヘッダを組み立てる。列は Mac 版・Android 版と共通。</summary>
    public static string BuildHeader(
        MEMEMode mode, MEMEQuality quality, MEMEAccelRange accelRange, MEMEGyroRange gyroRange)
    {
        var columns = mode switch
        {
            MEMEMode.Standard =>
                "//ARTIFACT,NUM,DATE,ACC_X,ACC_Y,ACC_Z,EOG_L1,EOG_R1,EOG_L2,EOG_R2,EOG_H1,EOG_H2,EOG_V1,EOG_V2",
            MEMEMode.Full =>
                "//ARTIFACT,NUM,DATE,ACC_X,ACC_Y,ACC_Z,GYRO_X,GYRO_Y,GYRO_Z,EOG_L,EOG_R,EOG_H,EOG_V",
            _ =>
                "//ARTIFACT,NUM,DATE,QUATERNION_W,QUATERNION_X,QUATERNION_Y,QUATERNION_Z",
        };

        return new StringBuilder()
            .Append("// Data mode  : ").Append(mode).Append("\r\n")
            .Append("// Transmission speed  : ").Append(quality == MEMEQuality.High ? "100Hz" : "50Hz").Append("\r\n")
            .Append("// Acceleration sensor's range  : ").Append(AccelRangeText(accelRange)).Append("\r\n")
            .Append("// Gyroscope sensor's range  : ").Append(GyroRangeText(gyroRange)).Append("\r\n")
            .Append("//\r\n")
            .Append(columns).Append("\r\n")
            .ToString();
    }

    /// <summary>CSV / TCP 共通の 1 行整形。DATE 列は UTC。</summary>
    public static string FormatRow(AcademicData data, int packetCount, bool freeMarking)
    {
        var mark = freeMarking ? "x" : string.Empty;
        var timestamp = (data.RecordedUtc ?? DateTime.UtcNow).ToString(DateFormat, CultureInfo.InvariantCulture);

        return data switch
        {
            AcademicStandardData d => string.Join(',',
                mark, packetCount, timestamp,
                d.AccX, d.AccY, d.AccZ,
                d.EogL1, d.EogR1, d.EogL2, d.EogR2,
                d.EogH1, d.EogH2, d.EogV1, d.EogV2),

            AcademicFullData d => string.Join(',',
                mark, packetCount, timestamp,
                d.AccX, d.AccY, d.AccZ,
                d.GyroX, d.GyroY, d.GyroZ,
                d.EogL, d.EogR, d.EogH, d.EogV),

            AcademicQuaternionData d => string.Join(',',
                mark, packetCount, timestamp,
                d.QuaternionW, d.QuaternionX, d.QuaternionY, d.QuaternionZ),

            _ => string.Join(',', mark, packetCount, timestamp),
        };
    }

    public static string AccelRangeText(MEMEAccelRange range) => range switch
    {
        MEMEAccelRange.Range2G => "2g",
        MEMEAccelRange.Range4G => "4g",
        MEMEAccelRange.Range8G => "8g",
        _ => "16g",
    };

    public static string GyroRangeText(MEMEGyroRange range) => range switch
    {
        MEMEGyroRange.Range250dps => "250dps",
        MEMEGyroRange.Range500dps => "500dps",
        MEMEGyroRange.Range1000dps => "1000dps",
        _ => "2000dps",
    };

    public void Dispose() => End();
}

File: Mac/ES_R_DevKit_Mac/MEME_Academic/Services/DataPersistenceService.swift
//
//  DataPersistenceService.swift
//  MEME_Academic
//
//  CSV ヘッダー生成・行整形・ファイル保存／移動を担当するサービス。
//

import Foundation
import AppKit
import UniformTypeIdentifiers

@MainActor
final class DataPersistenceService {

    private var csvManager = CsvManager()
    private var pendingCsvRows: [[String: Any]] = []

    // MARK: - Reset

    func reset() {
        csvManager = CsvManager()
        pendingCsvRows.removeAll()
    }

    func resetCsvManager() {
        csvManager.reset()
    }

    // MARK: - Buffer

    func append(_ row: [String: Any]) {
        pendingCsvRows.append(row)
    }

    var lastRow: [String: Any]? { pendingCsvRows.last }

    /// 現在書き出し中のCSVファイルURL。まだ1件も書き出していなければ nil。
    /// 計測停止時に、タップで付けた Artifact を確定済みファイルへ書き戻すために使う。
    var savedFileURL: URL? {
        guard let path = csvManager.saveFilePath, !path.isEmpty else { return nil }
        return URL(fileURLWithPath: path)
    }

    // MARK: - Save trigger

    /// バッファ件数が閾値を超える or force=true のとき書き出す。
    /// 初回書き込み時は header を含めたうえで保存する。
    func saveIfNeeded(force: Bool,
                      macAddress: String,
                      quality: Int,
                      mode: UInt32,
                      header: @autoclosure () -> String) {
        if pendingCsvRows.isEmpty { return }
        if force || pendingCsvRows.count >= 100 / quality {
            if !csvManager.isSave {
                let directoryPath = UserSetting.getSaveFilePath()
                // ファイル名の日時もUTC（DATE列・Android版 DevKit と揃える）。
                let formatter = DateFormatter()
                formatter.locale = Locale(identifier: "en_US_POSIX")
                formatter.timeZone = TimeZone(secondsFromGMT: 0)
                formatter.dateFormat = "yyyyMMddHHmmss"
                let dateString = formatter.string(from: Date())
                // 拡張子で圧縮の有無が決まる（CsvManager が .csv.gz なら gzip で書く）。
                // 読み込み側は設定に関係なく .csv / .csv.gz の両方を受け付ける。
                let ext = CsvFile.saveExtension(compressed: UserSetting.getCompressSaveFile())
                let fileName = "\(macAddress)_\(dateString).\(ext)"
                var buffer = header()
                dataToStoring(pendingCsvRows, stringBuffer: &buffer, mode: mode)
                if let data = buffer.data(using: .utf8) {
                    csvManager.create(directoryPath: directoryPath, fileName: fileName, firstData: data)
                }
            } else {
                var buffer = ""
                dataToStoring(pendingCsvRows, stringBuffer: &buffer, mode: mode)
                if let data = buffer.data(using: .utf8) {
                    csvManager.append(data)
                }
            }
            pendingCsvRows.removeAll()
        }
    }

    // MARK: - Header / row formatting

    /// 計測パラメータから CSV ヘッダ文字列を生成する。
    static func headerString(mode: UInt32, transMode: UInt32, accelRange: UInt32, gyroRange: UInt32) -> String {
        let selectModeStr: String = {
            switch mode {
            case MEMEMode_Standard: return "Standard"
            case MEMEMode_Full: return "Full"
            default: return "Quaternion"
            }
        }()
        let transModeStr = transMode == MEMEQuality_High ? "100Hz" : "50Hz"
        let accelRangeStr: String = {
            switch accelRange {
            case MEMEAccelRange_2G: return "2g"
            case MEMEAccelRange_4G: return "4g"
            case MEMEAccelRange_8G: return "8g"
            default: return "16g"
            }
        }()
        let gyroRangeStr: String = {
            switch gyroRange {
            case MEMEGyroRange_250dps: return "250dps"
            case MEMEGyroRange_500dps: return "500dps"
            case MEMEGyroRange_1000dps: return "1000dps"
            default: return "2000dps"
            }
        }()
        var s = ""
        s += "// Data mode  : \(selectModeStr)\n"
        s += "// Transmission speed  : \(transModeStr)\n"
        s += "// Acceleration sensor's range  : \(accelRangeStr)\n"
        s += "// Gyroscope sensor's range  : \(gyroRangeStr)\n"
        s += "//\n"
        switch mode {
        case MEMEMode_Standard:
            s += "//ARTIFACT,NUM,DATE,ACC_X,ACC_Y,ACC_Z,EOG_L1,EOG_R1,EOG_L2,EOG_R2,EOG_H1,EOG_H2,EOG_V1,EOG_V2\n"
        case MEMEMode_Full:
            s += "//ARTIFACT,NUM,DATE,ACC_X,ACC_Y,ACC_Z,GYRO_X,GYRO_Y,GYRO_Z,EOG_L,EOG_R,EOG_H,EOG_V\n"
        default:
            s += "//ARTIFACT,NUM,DATE,QUATERNION_W,QUATERNION_X,QUATERNION_Y,QUATERNION_Z\n"
        }
        return s
    }

    /// CSV/Socket 共通の1行整形ロジック。
    /// DATE列はUTCで書き出す（Android版 DevKit と同じ ES_R CSV フォーマット）。
    /// 表示側（チャートX軸）が Setting の "Convert displayed time to local time" に従って変換する。
    func dataToStoring(_ datas: [[String: Any]], stringBuffer: inout String, mode: UInt32) {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy/MM/dd HH:mm:ss.SS"
        for dic in datas {
            let date = dic["date"] as? Date ?? Date()
            let dateString = formatter.string(from: date)
            let packetCount = (dic["packetCount"] as? NSNumber)?.intValue ?? 0
            let isFreeMarkingValue = (dic["isFreeMarking"] as? NSNumber)?.boolValue ?? false
            let mark = isFreeMarkingValue ? "x" : ""
            switch mode {
            case MEMEMode_Standard:
                if let d = dic["data"] as? AcademicStandardData {
                    stringBuffer += "\(mark),\(packetCount),\(dateString),\(d.accX),\(d.accY),\(d.accZ),\(d.eogL1),\(d.eogR1),\(d.eogL2),\(d.eogR2),\(d.eogH1),\(d.eogH2),\(d.eogV1),\(d.eogV2)\n"
                }
            case MEMEMode_Full:
                if let d = dic["data"] as? AcademicFullData {
                    stringBuffer += "\(mark),\(packetCount),\(dateString),\(d.accX),\(d.accY),\(d.accZ),\(d.gyroX),\(d.gyroY),\(d.gyroZ),\(d.eogL),\(d.eogR),\(d.eogH),\(d.eogV)\n"
                }
            default:
                if let d = dic["data"] as? AcademicQuaternionData {
                    stringBuffer += "\(mark),\(packetCount),\(dateString),\(d.quaternionW),\(d.quaternionX),\(d.quaternionY),\(d.quaternionZ)\n"
                }
            }
        }
    }

    // MARK: - File move (save dialog)

    func presentSaveDialog() {
        guard let sourceFilePath = csvManager.saveFilePath, !sourceFilePath.isEmpty else {
            csvManager.reset()
            return
        }
        let saveFileName = csvManager.saveFileName ?? ""

        let savePanel = NSSavePanel()
        savePanel.canCreateDirectories = true
        savePanel.showsTagField = false
        savePanel.isExtensionHidden = false
        if let type = CsvFile.contentType(forFileName: saveFileName) {
            savePanel.allowedContentTypes = [type]
        }
        savePanel.nameFieldStringValue = saveFileName
        savePanel.level = .modalPanel

        guard let window = NSApp.mainWindow else { return }
        savePanel.beginSheetModal(for: window) { result in
            if result == .OK, let url = savePanel.url {
                do {
                    try FileManager.default.copyItem(at: URL(fileURLWithPath: sourceFilePath), to: url)
                    try? FileManager.default.removeItem(at: URL(fileURLWithPath: sourceFilePath))
                } catch {
                    NSLog("コピー失敗:%@", error.localizedDescription)
                }
            }
        }
    }
}
